// nse-bse-api/dist-cjs/nse/constants/index.js
"use strict";
/**
 * Constants and configuration values for NSE API
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PRIME_URL = exports.DEFAULT_TIMEOUT = exports.DEFAULT_HEADERS = exports.DEFAULT_USER_AGENT = exports.ARCHIVE_URL = exports.BASE_URL = exports.OPTION_INDICES = exports.UDIFF_SWITCH_DATE = exports.FNO_IT = exports.FNO_FINNIFTY = exports.FNO_NIFTY = exports.FNO_BANK = exports.HOLIDAY_TRADING = exports.HOLIDAY_CLEARING = exports.SEGMENT_DEBT = exports.SEGMENT_MF = exports.SEGMENT_SME = exports.SEGMENT_EQUITY = exports.NSE_VERSION = void 0;
exports.NSE_VERSION = "0.1.0";
// Market segments
exports.SEGMENT_EQUITY = "equities";
exports.SEGMENT_SME = "sme";
exports.SEGMENT_MF = "mf";
exports.SEGMENT_DEBT = "debt";
// Holiday types
exports.HOLIDAY_CLEARING = "clearing";
exports.HOLIDAY_TRADING = "trading";
// F&O indices
exports.FNO_BANK = "banknifty";
exports.FNO_NIFTY = "nifty";
exports.FNO_FINNIFTY = "finnifty";
exports.FNO_IT = "niftyit";
// Important dates
exports.UDIFF_SWITCH_DATE = new Date("2024-07-08");
// Option indices array
exports.OPTION_INDICES = [
    exports.FNO_BANK,
    exports.FNO_NIFTY,
    exports.FNO_FINNIFTY,
    exports.FNO_IT,
];
// API URLs
exports.BASE_URL = "https://www.nseindia.com/api";
exports.ARCHIVE_URL = "https://nsearchives.nseindia.com";
// HTTP headers
exports.DEFAULT_USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; rv:109.0) Gecko/20100101 Firefox/118.0";
exports.DEFAULT_HEADERS = {
    "User-Agent": exports.DEFAULT_USER_AGENT,
    Accept: "*/*",
    "Accept-Language": "en-US,en;q=0.5",
    "Accept-Encoding": "gzip, deflate",
    Referer: "https://www.nseindia.com/get-quotes/equity?symbol=HDFCBANK",
};
// Default timeout
exports.DEFAULT_TIMEOUT = 15000;
// Cookie priming URL
exports.PRIME_URL = "https://www.nseindia.com/option-chain";
