// nse-bse-api/dist/nse/api/ipo-api.js
/**
 * IPO-related API methods for NSE
 */
import { BASE_URL } from "../constants/index.js";
import { formatDateDMY } from "../utils/date-formatter.js";
export class IpoApi {
    httpClient;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * List current IPOs
     */
    async listCurrentIPO() {
        return await this.httpClient.request(`${BASE_URL}/ipo-current-issue`);
    }
    /**
     * List upcoming IPOs
     */
    async listUpcomingIPO() {
        return await this.httpClient.request(`${BASE_URL}/all-upcoming-issues?category=ipo`);
    }
    /**
     * Get IPO details
     */
    async getIpoDetails(params) {
        const symbol = params.symbol.toUpperCase();
        const series = params.series ?? "EQ";
        return await this.httpClient.request(`${BASE_URL}/ipo-detail`, { symbol, series });
    }
    /**
     * List past IPOs
     */
    async listPastIPO(from_date, to_date) {
        const to = to_date ?? new Date();
        const from = from_date ?? new Date(to.getTime() - 90 * 86400000);
        if (to < from) {
            throw new Error("Argument `to_date` cannot be less than `from_date`");
        }
        return await this.httpClient.request(`${BASE_URL}/public-past-issues`, {
            from_date: formatDateDMY(from),
            to_date: formatDateDMY(to),
        });
    }
}
