// nse-bse-api/dist/bse/constants/index.js
/**
 * Collection of constants grouped into classes. Use it to pass arguments.
 */
export class CATEGORY {
    /** Corp Announcements category names */
    static AGM = 'AGM/EGM';
    static BOARD_MEETING = 'Board Meeting';
    static UPDATE = 'Company Update';
    static ACTION = 'Corp. Action';
    static INSIDER = 'Insider Trading / SAST';
    static NEW_LISTING = 'New Listing';
    static RESULT = 'Result';
    static OTHERS = 'Others';
}
export class SEGMENT {
    /** BSE Market segments */
    static EQUITY = 'Equity';
    static MF = 'MF';
    static PREFERENCE_SHARES = 'Preference Shares';
    static DEBENTURES_BONDS = 'Debentures and Bonds';
    static EQUITY_INSTITUTIONAL = 'Equity - Institutional Series';
    static COMMERCIAL_PAPERS = 'Commercial Papers';
}
export class STATUS {
    /** Current status of the Financial instrument */
    static ACTIVE = 'Active';
    static SUSPENDED = 'Suspended';
    static DELISTED = 'Delisted';
}
export class INDEX {
    /** BSE Market Indices */
    static SENSEX = 'S&P BSE SENSEX';
    static BSE100 = 'S&P BSE 100';
    static BSE100_ESG = 'S&P BSE 100 ESG INDEX';
    static BSE100_LARGECAP_TMC = 'S&P BSE 100 LARGECAP TMC INDEX';
    static BSE150_MIDCAP = 'S&P BSE 150 MIDCAP INDEX';
    static BSE200 = 'S&P BSE 200';
    static BSE250_LARGEMIDCAP = 'S&P BSE 250 LARGEMIDCAP INDEX';
    static BSE250_SMALLCAP = 'S&P BSE 250 SMALLCAP INDEX';
    static BSE400_MIDSMALLCAP = 'S&P BSE 400 MIDSMALLCAP INDEX';
    static BSE500 = 'S&P BSE 500';
    static ALLCAP = 'S&P BSE ALLCAP';
    static AUTO = 'S&P BSE AUTO';
    static BANKEX = 'S&P BSE BANKEX';
    static BHARAT_22 = 'S&P BSE BHARAT 22 INDEX';
    static CAPITAL_GOODS = 'S&P BSE CAPITAL GOODS';
    static CARBONEX = 'S&P BSE CARBONEX';
    static COMMODITIES = 'S&P BSE COMMODITIES';
    static CONSUMER_DISCRETIONARY = 'S&P BSE CONSUMER DISCRETIONARY';
    static CONSUMER_DURABLES = 'S&P BSE CONSUMER DURABLES';
    static CPSE = 'S&P BSE CPSE';
    static DIVERSIFIED_FINANCIALS_REVENUE_GROWTH = 'S&P BSE DIVERSIFIED FINANCIALS REVENUE GROWTH INDEX';
    static DIVIDEND_STABILITY = 'S&P BSE DIVIDEND STABILITY INDEX';
    static DOLLEX_100 = 'S&P BSE DOLLEX 100';
    static DOLLEX_200 = 'S&P BSE DOLLEX 200';
    static DOLLEX_30 = 'S&P BSE DOLLEX 30';
    static ENERGY = 'S&P BSE ENERGY';
    static ENHANCED_VALUE = 'S&P BSE ENHANCED VALUE INDEX';
    static FAST_MOVING_CONSUMER_GOODS = 'S&P BSE FAST MOVING CONSUMER GOODS';
    static FINANCIAL_SERVICES = 'S&P BSE FINANCIAL SERVICES';
    static GREENEX = 'S&P BSE GREENEX';
    static HEALTHCARE = 'S&P BSE HEALTHCARE';
    static INDIA_INFRASTRUCTURE = 'S&P BSE INDIA INFRASTRUCTURE INDEX';
    static INDIA_MANUFACTURING = 'S&P BSE INDIA MANUFACTURING INDEX';
    static INDUSTRIALS = 'S&P BSE INDUSTRIALS';
    static INFORMATION_TECHNOLOGY = 'S&P BSE INFORMATION TECHNOLOGY';
    static IPO = 'S&P BSE IPO';
    static LARGECAP = 'S&P BSE LARGECAP';
    static LOW_VOLATILITY = 'S&P BSE LOW VOLATILITY INDEX';
    static METAL = 'S&P BSE METAL';
    static MIDCAP = 'S&P BSE MIDCAP';
    static MIDCAP_SELECT = 'S&P BSE MIDCAP SELECT INDEX';
    static MOMENTUM = 'S&P BSE MOMENTUM INDEX';
    static OIL_GAS = 'S&P BSE OIL & GAS';
    static POWER = 'S&P BSE POWER';
    static PRIVATE_BANKS = 'S&P BSE PRIVATE BANKS INDEX';
    static PSU = 'S&P BSE PSU';
    static QUALITY = 'S&P BSE QUALITY INDEX ';
    static REALTY = 'S&P BSE REALTY';
    static SENSEX_50 = 'S&P BSE SENSEX 50';
    static SENSEX_NEXT_50 = 'S&P BSE SENSEX NEXT 50';
    static SERVICES = 'S&P BSE SERVICES';
    static SMALLCAP = 'S&P BSE SMALLCAP';
    static SMALLCAP_SELECT = 'S&P BSE SMALLCAP SELECT INDEX';
    static SME_IPO = 'S&P BSE SME IPO';
    static TECK = 'S&P BSE TECK';
    static TELECOMMUNICATION = 'S&P BSE TELECOMMUNICATION';
    static UTILITIES = 'S&P BSE UTILITIES';
}
export class PURPOSE {
    /** Corp Actions purpose code */
    static BONUS = 'P5';
    static BUYBACK = 'P6';
    static DIVIDEND = 'P9';
    static PREFERENCE_DIVIDEND = 'P10';
    static SPLIT = 'P26';
    static DELISTING = 'P29';
}
export class SECTOR {
    /** BSE Sector names */
    static AUTO = 'Automobile and Auto Components';
    static CAPITAL_GOODS = 'Capital Goods';
    static CHEMICALS = 'Chemicals';
    static CONSTRUCTION = 'Construction';
    static CONSTRUCTION_MATERIALS = 'Construction Materials';
    static CONSUMER_DURABLES = 'Consumer Durables';
    static CONSUMER_SERVICES = 'Consumer Services';
    static DIVERSIFIED = 'Diversified';
    static FMCG = 'Fast Moving Consumer Goods';
    static FIN_SERVICES = 'Financial Services';
    static FOREST_MATERIALS = 'Forest Materials';
    static IT = 'Information Technology';
    static MEDIA = 'Media, Entertainment & Publication';
    static METAL_MINING = 'Metals & Mining';
    static OIL_GAS = 'Oil, Gas & Consumable Fuels';
    static POWER = 'Power';
    static REALTY = 'Realty';
    static SERVICES = 'Services';
    static TELECOM = 'Telecommunication';
    static TEXTILES = 'Textiles';
    static UTILITIES = 'Utilities';
}
export class SECTOR_ISIN {
    /** ISIN code for BSE Sectors */
    static AUTOMOBILE = 'IN020102001';
    static CAPITAL_GOODS = 'IN070205008';
    static CHEMICALS = 'IN010101001';
    static CONSTRUCTION = 'IN070101001';
    static CONSTRUCTION_MATERIALS = 'IN010203001';
    static CONSUMER_DURABLES = 'IN020201003';
    static CONSUMER_SERVICES = 'IN020601001';
    static DIVERSIFIED = 'IN120101001';
    static FMCG = 'IN040101003';
    static FIN_SERVICES = 'IN050103006';
    static FOREST_MATERIALS = 'IN010401001';
    static HEALTHCARE = 'IN060101001';
    static IT = 'IN080102001';
    static MEDIA = 'IN020401001';
    static METALS = 'IN010301001';
    static OIL_GAS = 'IN030103001';
    static POWER = 'IN110101002';
    static REALTY = 'IN020501001';
    static SERVICES = 'IN090104002';
    static TELECOMMUNICATION = 'IN100101003';
    static TEXTILES = 'IN020301002';
    static UTILITIES = 'IN110201005';
}
export class INDUSTRY {
    /** BSE Industry names */
    static TWO_WHEELERS = '2/3 WHEELERS';
    static ABRASIVES_BEARINGS = 'ABRASIVES & BEARINGS';
    static ADVERTISING = 'ADVERTISING & MEDIA AGENCIES';
    static AEROSPACE_DEFENSE = 'AEROSPACE & DEFENSE';
    static AIRLINE = 'AIRLINE';
    static AIRPORT_SERVICES = 'AIRPORT & AIRPORT SERVICES';
    static ALUMINIUM = 'ALUMINIUM';
    static ALUMINIUM_COPPER_ZINC = 'ALUMINIUM, COPPER & ZINC PRODUCTS';
    static AMUSEMENT_PARKS = 'AMUSEMENT PARKS/ OTHER RECREATION';
    static ANIMAL_FEED = 'ANIMAL FEED';
    static ASSET_MANAGEMENT = 'ASSET MANAGEMENT COMPANY';
    static AUTO_COMPONENTS = 'AUTO COMPONENTS & EQUIPMENTS';
    static AUTO_DEALER = 'AUTO -DEALER';
    static BIOTECHNOLOGY = 'BIOTECHNOLOGY';
    static BREWERIES = 'BREWERIES & DISTILLERIES';
    static BPO_KPO = 'BUSINESS PROCESS OUTSOURCING (BPO)/ KNOWLEDGE PROCESS OUTSOURCING (KPO)';
    static CABLES_ELECTRICALS = 'CABLES - ELECTRICALS';
    static CARBON_BLACK = 'CARBON BLACK';
    static CASTINGS_FORGINGS = 'CASTINGS & FORGINGS';
    static CEMENT = 'CEMENT & CEMENT PRODUCTS';
    static CERAMICS = 'CERAMICS';
    static TOBACCO = 'CIGARETTES & TOBACCO PRODUCTS';
    static CIVIL_CONSTRUCTION = 'CIVIL CONSTRUCTION';
    static COAL = 'COAL';
    static COMMERCIAL_VEHICLES = 'COMMERCIAL VEHICLES';
    static COMMODITY_CHEMICALS = 'COMMODITY CHEMICALS';
    static COMPRESSORS_PUMPS_ENGINES = 'COMPRESSORS, PUMPS & DIESEL ENGINES';
    static IT_SOFTWARE = 'COMPUTERS - SOFTWARE & CONSULTING';
    static IT_HARDWARE = 'COMPUTERS HARDWARE & EQUIPMENTS';
    static CONSTRUCTION_VEHICLES = 'CONSTRUCTION VEHICLES';
    static CONSULTING_SERVICES = 'CONSULTING SERVICES';
    static CONSUMER_ELECTRONICS = 'CONSUMER ELECTRONICS';
    static COPPER = 'COPPER';
    static CYCLES = 'CYCLES';
    static DAIRY_PRODUCTS = 'DAIRY PRODUCTS';
    static DATA_PROCESSING = 'DATA PROCESSING SERVICES';
    static DEPOSITORIES_CLEARING_HOUSES = 'DEPOSITORIES, CLEARING HOUSES AND OTHER INTERMEDIARIES';
    static DIGITAL_ENTERTAINMENT = 'DIGITAL ENTERTAINMENT';
    static DISTRIBUTORS = 'DISTRIBUTORS';
    static DIVERSIFIED = 'DIVERSIFIED';
    static DIVERSIFIED_COMMERCIAL = 'DIVERSIFIED COMMERCIAL SERVICES';
    static DIVERSIFIED_CONSUMER = 'DIVERSIFIED CONSUMER PRODUCTS';
    static DIVERSIFIED_FMCG = 'DIVERSIFIED FMCG';
    static DIVERSIFIED_METALS = 'DIVERSIFIED METALS';
    static DIVERSIFIED_RETAIL = 'DIVERSIFIED RETAIL';
    static DREDGING = 'DREDGING';
    static DYES_AND_PIGMENTS = 'DYES AND PIGMENTS';
    static EDIBLE_OIL = 'EDIBLE OIL';
    static EDUCATION = 'EDUCATION';
    static ELEARNING = 'E-LEARNING';
    static ELECTRODES_REFRACTORIES = 'ELECTRODES & REFRACTORIES';
    static ELECTRONIC_MEDIA = 'ELECTRONIC MEDIA';
    static ECOMMERCE = 'E-RETAIL/ E-COMMERCE';
    static EXCHANGE_PLATFORM = 'EXCHANGE AND DATA PLATFORM';
    static EXPLOSIVES = 'EXPLOSIVES';
    static FERRO_SILICA_MANGANESE = 'FERRO & SILICA MANGANESE';
    static FERTILIZERS = 'FERTILIZERS';
    static FILM_PRODUCTION = 'FILM PRODUCTION, DISTRIBUTION & EXHIBITION';
    static FINANCIAL_INSTITUTION = 'FINANCIAL INSTITUTION';
    static FINANCIAL_PRODUCTS = 'FINANCIAL PRODUCTS DISTRIBUTOR';
    static FINTECH = 'FINANCIAL TECHNOLOGY (FINTECH)';
    static FOOD_STORAGE = 'FOOD STORAGE FACILITIES';
    static FOOTWEAR = 'FOOTWEAR';
    static FOREST_PRODUCTS = 'FOREST PRODUCTS';
    static FURNITURE_HOME_FURNISHING = 'FURNITURE, HOME FURNISHING';
    static FURNITURE_HOME_FURNISHING_FLOORING = 'FURNITURE, HOME FURNISHING, FLOORING';
    static GARMENTS_APPARELS = 'GARMENTS & APPARELS';
    static GAS_TRANSMISSION = 'GAS TRANSMISSION/MARKETING';
    static JEWELLERY = 'GEMS, JEWELLERY AND WATCHES';
    static GENERAL_INSURANCE = 'GENERAL INSURANCE';
    static GLASS_CONSUMER = 'GLASS - CONSUMER';
    static GLASS_INDUSTRIAL = 'GLASS - INDUSTRIAL';
    static GRANITES_MARBLES = 'GRANITES & MARBLES';
    static HEALTHCARE = 'HEALTHCARE RESEARCH, ANALYTICS & TECHNOLOGY';
    static HEALTHCARE_SERVICE = 'HEALTHCARE SERVICE PROVIDER';
    static HEAVY_ELECTRICAL = 'HEAVY ELECTRICAL EQUIPMENT';
    static HOLDING_COMPANY = 'HOLDING COMPANY';
    static HOSPITAL = 'HOSPITAL';
    static HOTELS_RESORTS = 'HOTELS & RESORTS';
    static HOUSEHOLD_APPLIANCES = 'HOUSEHOLD APPLIANCES';
    static HOUSEHOLD_PRODUCTS = 'HOUSEHOLD PRODUCTS';
    static HOUSEWARE = 'HOUSEWARE';
    static HOUSING_FINANCE_COMPANY = 'HOUSING FINANCE COMPANY';
    static INDUSTRIAL_GASES = 'INDUSTRIAL GASES';
    static INDUSTRIAL_MACHINERY = 'INDUSTRIAL MACHINERY';
    static INDUSTRIAL_MINERALS = 'INDUSTRIAL MINERALS';
    static INDUSTRIAL_PRODUCTS = 'INDUSTRIAL PRODUCTS';
    static INTEGRATED_POWER_UTILITIES = 'INTEGRATED POWER UTILITIES';
    static INTERNET_CATALOGUE_RETAIL = 'INTERNET & CATALOGUE RETAIL';
    static INVESTMENT_COMPANY = 'INVESTMENT COMPANY';
    static IRON_STEEL = 'IRON & STEEL';
    static IRON_STEEL_PRODUCTS = 'IRON & STEEL PRODUCTS';
    static IT_ENABLED_SERVICES = 'IT ENABLED SERVICES';
    static JUTE_JUTE_PRODUCTS = 'JUTE & JUTE PRODUCTS';
    static LEATHER_AND_LEATHER_PRODUCTS = 'LEATHER AND LEATHER PRODUCTS';
    static LEISURE_PRODUCTS = 'LEISURE PRODUCTS';
    static LIFE_INSURANCE = 'LIFE INSURANCE';
    static LOGISTICS_SOLUTION_PROVIDER = 'LOGISTICS SOLUTION PROVIDER';
    static LPG_CNG_PNG_LNG = 'LPG/CNG/PNG/LNG SUPPLIER';
    static LUBRICANTS = 'LUBRICANTS';
    static MEAT_PRODUCTS_INCLUDING_POULTRY = 'MEAT PRODUCTS INCLUDING POULTRY';
    static MEDIA_ENTERTAINMENT = 'MEDIA & ENTERTAINMENT';
    static MEDICAL_EQUIPMENT_SUPPLIES = 'MEDICAL EQUIPMENT & SUPPLIES';
    static MICROFINANCE_INSTITUTIONS = 'MICROFINANCE INSTITUTIONS';
    static MUTUAL_FUND = 'MUTUAL FUND SCHEME - ETF';
    static NBFC = 'NON BANKING FINANCIAL COMPANY (NBFC)';
    static OFFSHORE_DRILLING = 'OFFSHORE SUPPORT SOLUTION DRILLING';
    static OIL_EQUIPMENT_SERVICES = 'OIL EQUIPMENT & SERVICES';
    static OIL_EXPLORATION_PRODUCTION = 'OIL EXPLORATION & PRODUCTION';
    static OIL_STORAGE_TRANSPORT = 'OIL STORAGE & TRANSPORTATION';
    static OTHER_AGRICULTURAL_PRODUCTS = 'OTHER AGRICULTURAL PRODUCTS';
    static OTHER_BANK = 'OTHER BANK';
    static OTHER_BEVERAGES = 'OTHER BEVERAGES';
    static OTHER_CAPITAL_MARKET_SERVICES = 'OTHER CAPITAL MARKET RELATED SERVICES';
    static OTHER_CONSTRUCTION_MATERIALS = 'OTHER CONSTRUCTION MATERIALS';
    static OTHER_CONSUMER_SERVICES = 'OTHER CONSUMER SERVICES';
    static OTHER_ELECTRICAL_EQUIPMENT = 'OTHER ELECTRICAL EQUIPMENT';
    static OTHER_FINANCIAL_SERVICES = 'OTHER FINANCIAL SERVICES';
    static OTHER_FOOD_PRODUCTS = 'OTHER FOOD PRODUCTS';
    static OTHER_INDUSTRIAL_PRODUCTS = 'OTHER INDUSTRIAL PRODUCTS';
    static OTHER_TELECOM_SERVICES = 'OTHER TELECOM SERVICES';
    static OTHER_TEXTILE_PRODUCTS = 'OTHER TEXTILE PRODUCTS';
    static OTHER_UTILITIES = 'OTHER UTILITIES';
    static PACKAGED_FOODS = 'PACKAGED FOODS';
    static PACKAGING = 'PACKAGING';
    static PAINTS = 'PAINTS';
    static PAPER_PAPER_PRODUCTS = 'PAPER & PAPER PRODUCTS';
    static PASSENGER_CARS_UTILITY_VEHICLES = 'PASSENGER CARS & UTILITY VEHICLES';
    static PERSONAL_CARE = 'PERSONAL CARE';
    static PESTICIDES_AGROCHEMICALS = 'PESTICIDES & AGROCHEMICALS';
    static PETROCHEMICALS = 'PETROCHEMICALS';
    static PHARMACEUTICALS = 'PHARMACEUTICALS';
    static PHARMACY_RETAIL = 'PHARMACY RETAIL';
    static PHOTOGRAPHIC_PRODUCTS = 'PHOTOGRAPHIC PRODUCTS';
    static PIG_IRON = 'PIG IRON';
    static PLASTIC_PRODUCTS_CONSUMER = 'PLASTIC PRODUCTS - CONSUMER';
    static PLASTIC_PRODUCTS_INDUSTRIAL = 'PLASTIC PRODUCTS - INDUSTRIAL';
    static PLYWOOD_BOARDS_LAMINATES = 'PLYWOOD BOARDS/ LAMINATES';
    static PORT_SERVICES = 'PORT & PORT SERVICES';
    static POWER_TRANSMISSION = 'POWER - TRANSMISSION';
    static POWER_GENERATION = 'POWER GENERATION';
    static POWER_TRADING = 'POWER TRADING';
    static PRECIOUS_METALS = 'PRECIOUS METALS';
    static PRINT_MEDIA = 'PRINT MEDIA';
    static PRINTING_PUBLICATION = 'PRINTING & PUBLICATION';
    static PRINTING_INKS = 'PRINTING INKS';
    static PRIVATE_SECTOR_BANK = 'PRIVATE SECTOR BANK';
    static PUBLIC_SECTOR_BANK = 'PUBLIC SECTOR BANK';
    static RAILWAY_WANS = 'RAILWAY WANS';
    static RATINGS = 'RATINGS';
    static REITS = 'REAL ESTATE INVESTMENT TRUSTS (REITS)';
    static REAL_ESTATE = 'REAL ESTATE RELATED SERVICES';
    static REFINERIES_MARKETING = 'REFINERIES & MARKETING';
    static RESIDENTIAL_COMMERCIAL_PROJECTS = 'RESIDENTIAL, COMMERCIAL PROJECTS';
    static RESTAURANTS = 'RESTAURANTS';
    static ROAD_ASSETS = 'ROAD ASSETS–TOLL, ANNUITY, HYBRID-ANNUITY';
    static ROAD_TRANSPORT = 'ROAD TRANSPORT';
    static RUBBER = 'RUBBER';
    static SANITARY_WARE = 'SANITARY WARE';
    static SEAFOOD = 'SEAFOOD';
    static SHIP_BUILDING = 'SHIP BUILDING & ALLIED SERVICES';
    static SHIPPING = 'SHIPPING';
    static SOFTWARE_PRODUCTS = 'SOFTWARE PRODUCTS';
    static SPECIALITY_RETAIL = 'SPECIALITY RETAIL';
    static SPECIALTY_CHEMICALS = 'SPECIALTY CHEMICALS';
    static SPONGE_IRON = 'SPONGE IRON';
    static STATIONARY = 'STATIONARY';
    static STOCKBROKING = 'STOCKBROKING & ALLIED';
    static SUGAR = 'SUGAR';
    static TEA_COFFEE = 'TEA & COFFEE';
    static TELECOM_EQUIPMENT = 'TELECOM -  EQUIPMENT & ACCESSORIES';
    static TELECOM_CELLULAR = 'TELECOM - CELLULAR & FIXED LINE SERVICES';
    static TELECOM_INFRASTRUCTURE = 'TELECOM - INFRASTRUCTURE';
    static TOUR_TRAVEL = 'TOUR, TRAVEL RELATED SERVICES';
    static TRACTORS = 'TRACTORS';
    static TRADING_AUTO_COMPONENTS = 'TRADING - AUTO COMPONENTS';
    static TRADING_CHEMICALS = 'TRADING - CHEMICALS';
    static TRADING_GAS = 'TRADING - GAS';
    static TRADING_METALS = 'TRADING - METALS';
    static TRADING_MINERALS = 'TRADING - MINERALS';
    static TRADING_TEXTILE = 'TRADING - TEXTILE PRODUCTS';
    static TRADING_DISTRIBUTORS = 'TRADING & DISTRIBUTORS';
    static TRADING_COAL = 'TRADING COAL';
    static TRANSPORT_RELATED_SERVICES = 'TRANSPORT RELATED SERVICES';
    static TV_BROADCASTING = 'TV BROADCASTING & SOFTWARE PRODUCTION';
    static TYRES_RUBBER = 'TYRES & RUBBER PRODUCTS';
    static WASTE_MANAGEMENT = 'WASTE MANAGEMENT';
    static WATER_SUPPLY = 'WATER SUPPLY & MANAGEMENT';
    static WELLNESS = 'WELLNESS';
    static ZINC = 'ZINC';
}
