// nse-bse-api/dist-cjs/bse/utils/helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPath = getPath;
exports.splitDateRange = splitDateRange;
exports.formatDate = formatDate;
exports.formatDateSlash = formatDateSlash;
exports.unzipFile = unzipFile;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const index_js_1 = require("../types/index.js");
/**
 * Utility function to get and validate path
 */
function getPath(inputPath, isFolder = false) {
    const resolvedPath = path.resolve(inputPath);
    if (isFolder) {
        if (fs.existsSync(resolvedPath) && fs.statSync(resolvedPath).isFile()) {
            throw new index_js_1.BSEError(`${resolvedPath}: must be a folder`);
        }
        if (!fs.existsSync(resolvedPath)) {
            fs.mkdirSync(resolvedPath, { recursive: true });
        }
    }
    return resolvedPath;
}
/**
 * Split date range into chunks
 */
function splitDateRange(fromDate, toDate, maxChunkSize = 30) {
    if (fromDate > toDate) {
        throw new index_js_1.BSEError("'fromDate' cannot be greater than 'toDate'");
    }
    const chunks = [];
    let currentStart = new Date(fromDate);
    while (currentStart <= toDate) {
        // Calculate the end of the current chunk
        const currentEnd = new Date(currentStart);
        currentEnd.setDate(currentEnd.getDate() + maxChunkSize - 1);
        // Don't go past the final date
        if (currentEnd > toDate) {
            currentEnd.setTime(toDate.getTime());
        }
        chunks.push([new Date(currentStart), new Date(currentEnd)]);
        // Start next chunk the day after the current end
        currentStart = new Date(currentEnd);
        currentStart.setDate(currentStart.getDate() + 1);
    }
    return chunks;
}
/**
 * Format date to YYYYMMDD string
 */
function formatDate(date) {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}${month}${day}`;
}
/**
 * Format date to DD/MM/YYYY string
 */
function formatDateSlash(date) {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${day}/${month}/${year}`;
}
/**
 * Extract and unzip file
 */
async function unzipFile(filePath, extractPath) {
    const AdmZip = require('adm-zip');
    const zip = new AdmZip(filePath);
    const zipEntries = zip.getEntries();
    if (zipEntries.length === 0) {
        throw new index_js_1.BSEError('Empty zip file');
    }
    const extractedPath = path.join(extractPath, zipEntries[0].entryName);
    zip.extractAllTo(extractPath, true);
    // Remove the original zip file
    fs.unlinkSync(filePath);
    return extractedPath;
}
