// nse-bse-api/dist-cjs/bse/utils/SymbolParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SymbolParser = void 0;
const node_html_parser_1 = require("node-html-parser");
/**
 * HTML parser to parse html strings returned from BSE symbol search
 * The result is parsed into a dictionary with company_name, symbol, isin and bse_code
 */
class SymbolParser {
    result = {};
    start = false;
    index = 0;
    // The fields within the HTML are always in the same order
    fields = ['company_name', 'symbol', 'isin', 'bse_code'];
    resetData() {
        this.result = {};
        this.start = false;
        this.index = 0;
    }
    getResult() {
        if (!this.result.symbol) {
            return null;
        }
        return this.result;
    }
    feed(htmlString) {
        const root = (0, node_html_parser_1.parse)(htmlString);
        const anchors = root.querySelectorAll('a');
        if (anchors.length === 0) {
            return;
        }
        this.start = true;
        this.index = 0;
        // Process each anchor tag's text content
        for (const anchor of anchors) {
            const textContent = anchor.text.trim();
            if (textContent) {
                this.handleData(textContent);
            }
        }
    }
    handleData(data) {
        if (!this.start || this.index >= this.fields.length) {
            return;
        }
        const field = this.fields[this.index];
        if (!(field in this.result)) {
            if ('company_name' in this.result && data.includes(' ')) {
                // Handle strings like "   INE040A01034   500180"
                const parts = data.split(' ').filter(part => part.trim() !== '');
                for (const part of parts) {
                    this.handleData(part);
                }
                return;
            }
            this.result[field] = data.trim();
            this.index++;
        }
    }
}
exports.SymbolParser = SymbolParser;
