// nse-bse-api/dist-cjs/nse/nse/nse-client.js
"use strict";
/**
 * Main NSE API client that orchestrates all API modules
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NSEClient = void 0;
const http_client_js_1 = require("../http/http-client.js");
const equity_api_js_1 = require("../api/equity-api.js");
const options_api_js_1 = require("../api/options-api.js");
const historical_api_js_1 = require("../api/historical-api.js");
const corporate_api_js_1 = require("../api/corporate-api.js");
const ipo_api_js_1 = require("../api/ipo-api.js");
const market_api_js_1 = require("../api/market-api.js");
const download_api_js_1 = require("../api/download-api.js");
const file_operations_js_1 = require("../utils/file-operations.js");
const index_js_1 = require("../constants/index.js");
class NSEClient {
    static __version__ = "0.1.0";
    // Re-export constants for backward compatibility
    static SEGMENT_EQUITY = "equities";
    static SEGMENT_SME = "sme";
    static SEGMENT_MF = "mf";
    static SEGMENT_DEBT = "debt";
    static HOLIDAY_CLEARING = "clearing";
    static HOLIDAY_TRADING = "trading";
    static FNO_BANK = "banknifty";
    static FNO_NIFTY = "nifty";
    static FNO_FINNIFTY = "finnifty";
    static FNO_IT = "niftyit";
    static UDIFF_SWITCH_DATE = new Date("2024-07-08");
    httpClient;
    downloadDir;
    // API modules
    equity;
    options;
    historical;
    corporate;
    ipo;
    market;
    download;
    constructor(downloadFolder, opts = {}) {
        const server = !!opts.server;
        const timeout = opts.timeout ?? index_js_1.DEFAULT_TIMEOUT;
        this.downloadDir = (0, file_operations_js_1.getPath)(downloadFolder, true);
        this.httpClient = new http_client_js_1.HttpClient(this.downloadDir, server, timeout);
        // Initialize API modules
        this.equity = new equity_api_js_1.EquityApi(this.httpClient);
        this.options = new options_api_js_1.OptionsApi(this.httpClient);
        this.historical = new historical_api_js_1.HistoricalApi(this.httpClient);
        this.corporate = new corporate_api_js_1.CorporateApi(this.httpClient);
        this.ipo = new ipo_api_js_1.IpoApi(this.httpClient);
        this.market = new market_api_js_1.MarketApi(this.httpClient);
        this.download = new download_api_js_1.DownloadApi(this.httpClient, this.downloadDir);
    }
    /**
     * Clean up resources and save cookies
     */
    exit() {
        this.httpClient.saveCookies();
    }
    // Backward compatibility methods - delegate to appropriate modules
    async status() { return this.market.getStatus(); }
    async lookup(query) { return this.market.lookup(query); }
    async equityBhavcopy(date, folder) { return this.download.downloadEquityBhavcopy(date, folder); }
    async deliveryBhavcopy(date, folder) { return this.download.downloadDeliveryBhavcopy(date, folder); }
    async indicesBhavcopy(date, folder) { return this.download.downloadIndicesBhavcopy(date, folder); }
    async fnoBhavcopy(date, folder) { return this.download.downloadFnoBhavcopy(date, folder); }
    async priceband_report(date, folder) { return this.download.downloadPricebandReport(date, folder); }
    async pr_bhavcopy(date, folder) { return this.download.downloadPrBhavcopy(date, folder); }
    async cm_mii_security_report(date, folder) { return this.download.downloadCmMiiSecurityReport(date, folder); }
    async download_document(url, folder, extract_files) { return this.download.downloadDocument(url, folder, extract_files); }
    async actions(params = {}) { return this.corporate.getActions(params); }
    async announcements(params = {}) { return this.corporate.getAnnouncements(params); }
    async boardMeetings(params = {}) { return this.corporate.getBoardMeetings(params); }
    async annual_reports(symbol, segment = "equities") { return this.corporate.getAnnualReports(symbol, segment); }
    async circulars(params = {}) { return this.corporate.getCirculars(params); }
    async equityMetaInfo(symbol) { return this.equity.getEquityMetaInfo(symbol); }
    async quote(params) { return this.equity.getQuote(params); }
    async equityQuote(symbol) { return this.equity.getEquityQuote(symbol); }
    gainers(data, count) { return this.equity.getGainers(data, count); }
    losers(data, count) { return this.equity.getLosers(data, count); }
    async listEquityStocksByIndex(index) { return this.equity.listEquityStocksByIndex(index); }
    async listIndices() { return this.equity.listIndices(); }
    async listEtf() { return this.equity.listEtf(); }
    async listSme() { return this.equity.listSme(); }
    async listSgb() { return this.equity.listSgb(); }
    async listCurrentIPO() { return this.ipo.listCurrentIPO(); }
    async listUpcomingIPO() { return this.ipo.listUpcomingIPO(); }
    async getIpoDetails(params) { return this.ipo.getIpoDetails(params); }
    async listPastIPO(from_date, to_date) { return this.ipo.listPastIPO(from_date, to_date); }
    async blockDeals() { return this.market.getBlockDeals(); }
    async bulkdeals(fromdate, todate) { return this.market.getBulkDeals(fromdate, todate); }
    async holidays(type = "trading") { return this.market.getHolidays(type); }
    async fnoLots() { return this.options.getFnoLots(); }
    async optionChain(symbol) { return this.options.getOptionChain(symbol); }
    async filteredOptionChain(symbol, strikeRange) { return this.options.getFilteredOptionChain(symbol, strikeRange); }
    static maxpain(optionChain, expiryDate) { return options_api_js_1.OptionsApi.calculateMaxPain(optionChain, expiryDate); }
    async getFuturesExpiry(index = "nifty") { return this.options.getFuturesExpiry(index); }
    async compileOptionChain(symbol, expiryDate) { return this.options.compileOptionChain(symbol, expiryDate); }
    async fetch_equity_historical_data(params) { return this.historical.fetchEquityHistoricalData(params); }
    async fetch_historical_vix_data(params = {}) { return this.historical.fetchHistoricalVixData(params); }
    async fetch_historical_fno_data(params) { return this.historical.fetchHistoricalFnoData(params); }
    async fetch_historical_index_data(params) { return this.historical.fetchHistoricalIndexData(params); }
    async fetch_fno_underlying() { return this.historical.fetchFnoUnderlying(); }
    async fetch_index_names() { return this.historical.fetchIndexNames(); }
    async fetch_daily_reports_file_metadata(segment = "CM") { return this.market.getDailyReportsFileMetadata(segment); }
}
exports.NSEClient = NSEClient;
