// nse-bse-api/dist-cjs/nse/http/http-client.js
"use strict";
/**
 * HTTP client management for NSE API
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpClient = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const axios_1 = __importDefault(require("axios"));
const axios_cookiejar_support_1 = require("axios-cookiejar-support");
const got_1 = __importDefault(require("got"));
const tough_cookie_1 = require("tough-cookie");
const index_js_1 = require("../constants/index.js");
const file_operations_js_1 = require("../utils/file-operations.js");
class HttpClient {
    axiosClient;
    gotClient;
    jar;
    cookiePath;
    cookiesPrimed = false;
    server;
    timeout;
    constructor(downloadDir, server, timeout) {
        this.server = server;
        this.timeout = timeout;
        // Set up cookie jar and load from disk
        this.cookiePath = path_1.default.join(downloadDir, server ? "nse_cookies_http2.json" : "nse_cookies_http1.json");
        this.jar = new tough_cookie_1.CookieJar();
        this.loadCookies();
        if (server) {
            // got with HTTP/2
            this.gotClient = got_1.default.extend({
                http2: true,
                headers: index_js_1.DEFAULT_HEADERS,
                cookieJar: this.jar,
                timeout: { request: this.timeout },
                throwHttpErrors: false,
            });
        }
        else {
            // axios with cookie jar
            const ax = axios_1.default.create({
                headers: index_js_1.DEFAULT_HEADERS,
                timeout: this.timeout,
                decompress: true,
                validateStatus: () => true,
            });
            (0, axios_cookiejar_support_1.wrapper)(ax);
            ax.defaults.jar = this.jar;
            this.axiosClient = ax;
        }
    }
    /**
     * Load cookies from disk
     */
    loadCookies() {
        if (!fs_1.default.existsSync(this.cookiePath))
            return;
        try {
            const data = JSON.parse(fs_1.default.readFileSync(this.cookiePath, "utf8"));
            this.jar = tough_cookie_1.CookieJar.fromJSON(data);
        }
        catch {
            // Ignore cookie loading errors
        }
    }
    /**
     * Save cookies to disk
     */
    saveCookies() {
        try {
            const json = this.jar.toJSON();
            fs_1.default.writeFileSync(this.cookiePath, JSON.stringify(json), "utf8");
        }
        catch {
            // Ignore cookie saving errors
        }
    }
    /**
     * Prime cookies by visiting NSE homepage
     */
    async primeCookies() {
        if (this.cookiesPrimed)
            return;
        try {
            if (this.server && this.gotClient) {
                await this.gotClient.get(index_js_1.PRIME_URL);
            }
            else if (this.axiosClient) {
                await this.axiosClient.get(index_js_1.PRIME_URL);
            }
            this.cookiesPrimed = true;
            this.saveCookies();
        }
        catch {
            // Ignore priming failures; subsequent requests may still succeed
        }
    }
    /**
     * Make HTTP request and return JSON response
     */
    async request(url, params) {
        await this.primeCookies();
        if (this.server && this.gotClient) {
            const res = await this.gotClient.get(url, {
                searchParams: params,
            });
            if (res.statusCode < 200 || res.statusCode >= 300) {
                throw new Error(`${url} ${res.statusCode}: ${res.statusMessage}`);
            }
            return JSON.parse(res.body);
        }
        if (this.axiosClient) {
            const res = await this.axiosClient.get(url, { params });
            if (res.status < 200 || res.status >= 300) {
                throw new Error(`${url} ${res.status}: ${res.statusText}`);
            }
            return res.data;
        }
        throw new Error("HTTP client not initialized");
    }
    /**
     * Download file from URL
     */
    async downloadFile(url, folder) {
        const fname = path_1.default.join(folder, path_1.default.basename(new URL(url).pathname));
        await this.primeCookies();
        if (this.server && this.gotClient) {
            const res = await this.gotClient.get(url, { responseType: "buffer" });
            const contentType = res.headers["content-type"];
            if (contentType && String(contentType).includes("text/html")) {
                throw new Error("NSE file is unavailable or not yet updated.");
            }
            fs_1.default.writeFileSync(fname, res.body);
        }
        else if (this.axiosClient) {
            const res = await this.axiosClient.get(url, { responseType: "stream" });
            const contentType = res.headers["content-type"];
            if (contentType && contentType.includes("text/html")) {
                throw new Error("NSE file is unavailable or not yet updated.");
            }
            await (0, file_operations_js_1.saveStreamToFile)(res.data, fname);
        }
        else {
            throw new Error("HTTP client not initialized");
        }
        return fname;
    }
    /**
     * Get text response from URL
     */
    async getTextResponse(url) {
        await this.primeCookies();
        if (this.server && this.gotClient) {
            const res = await this.gotClient.get(url, { responseType: "text" });
            return res.body;
        }
        else if (this.axiosClient) {
            const res = await this.axiosClient.get(url, { responseType: "text" });
            return res.data;
        }
        else {
            throw new Error("HTTP client not initialized");
        }
    }
}
exports.HttpClient = HttpClient;
