// nse-bse-api/dist-cjs/nse/api/corporate-api.js
"use strict";
/**
 * Corporate actions and announcements API methods for NSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CorporateApi = void 0;
const index_js_1 = require("../constants/index.js");
const date_formatter_js_1 = require("../utils/date-formatter.js");
class CorporateApi {
    httpClient;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * Get corporate actions
     */
    async getActions(params = {}) {
        const query = {
            index: params.segment ?? "equities",
        };
        if (params.symbol)
            query.symbol = params.symbol;
        if (params.from_date && params.to_date) {
            (0, date_formatter_js_1.validateDateRange)(params.from_date, params.to_date);
            query.from_date = (0, date_formatter_js_1.formatDateDMY)(params.from_date);
            query.to_date = (0, date_formatter_js_1.formatDateDMY)(params.to_date);
        }
        return await this.httpClient.request(`${index_js_1.BASE_URL}/corporates-corporateActions`, query);
    }
    /**
     * Get corporate announcements
     */
    async getAnnouncements(params = {}) {
        const query = {
            index: params.index ?? "equities",
        };
        if (params.symbol)
            query.symbol = params.symbol;
        if (params.fno)
            query.fo_sec = true;
        if (params.from_date && params.to_date) {
            (0, date_formatter_js_1.validateDateRange)(params.from_date, params.to_date);
            query.from_date = (0, date_formatter_js_1.formatDateDMY)(params.from_date);
            query.to_date = (0, date_formatter_js_1.formatDateDMY)(params.to_date);
        }
        return await this.httpClient.request(`${index_js_1.BASE_URL}/corporate-announcements`, query);
    }
    /**
     * Get board meetings
     */
    async getBoardMeetings(params = {}) {
        const query = {
            index: params.index ?? "equities",
        };
        if (params.symbol)
            query.symbol = params.symbol;
        if (params.fno)
            query.fo_sec = true;
        if (params.from_date && params.to_date) {
            (0, date_formatter_js_1.validateDateRange)(params.from_date, params.to_date);
            query.from_date = (0, date_formatter_js_1.formatDateDMY)(params.from_date);
            query.to_date = (0, date_formatter_js_1.formatDateDMY)(params.to_date);
        }
        return await this.httpClient.request(`${index_js_1.BASE_URL}/corporate-board-meetings`, query);
    }
    /**
     * Get annual reports
     */
    async getAnnualReports(symbol, segment = "equities") {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/annual-reports`, {
            index: segment,
            symbol,
        });
    }
    /**
     * Get circulars
     */
    async getCirculars(params = {}) {
        const to = params.to_date ?? new Date();
        const from = params.from_date ?? new Date(to.getTime() - 7 * 86400000);
        (0, date_formatter_js_1.validateDateRange)(from, to);
        const query = {
            from_date: (0, date_formatter_js_1.formatDateDMY)(from),
            to_date: (0, date_formatter_js_1.formatDateDMY)(to),
        };
        if (params.subject)
            query.sub = params.subject;
        if (params.dept_code)
            query.dept = params.dept_code.toUpperCase();
        return await this.httpClient.request(`${index_js_1.BASE_URL}/circulars`, query);
    }
}
exports.CorporateApi = CorporateApi;
