// nse-bse-api/dist-cjs/nse/api/download-api.js
"use strict";
/**
 * File download and bhavcopy API methods for NSE
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadApi = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const index_js_1 = require("../constants/index.js");
const file_operations_js_1 = require("../utils/file-operations.js");
const date_formatter_js_1 = require("../utils/date-formatter.js");
class DownloadApi {
    httpClient;
    downloadDir;
    constructor(httpClient, downloadDir) {
        this.httpClient = httpClient;
        this.downloadDir = downloadDir;
    }
    /**
     * Download and extract equity bhavcopy
     */
    async downloadEquityBhavcopy(date, folder) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        let url;
        if (date < index_js_1.UDIFF_SWITCH_DATE) {
            const dateStr = (0, date_formatter_js_1.formatDateArchive)(date);
            const month = dateStr.substring(2, 5);
            url = `${index_js_1.ARCHIVE_URL}/content/historical/EQUITIES/${date.getFullYear()}/${month}/cm${dateStr}bhav.csv.zip`;
        }
        else {
            const ymd = (0, date_formatter_js_1.formatDateYMD)(date);
            url = `${index_js_1.ARCHIVE_URL}/content/cm/BhavCopy_NSE_CM_0_0_0_${ymd}_F_0000.csv.zip`;
        }
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        return await (0, file_operations_js_1.unzipFile)(file, path_1.default.dirname(file));
    }
    /**
     * Download delivery bhavcopy
     */
    async downloadDeliveryBhavcopy(date, folder) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        const ddmmyyyy = (0, date_formatter_js_1.formatDateDDMMYYYY)(date);
        const url = `${index_js_1.ARCHIVE_URL}/products/content/sec_bhavdata_full_${ddmmyyyy}.csv`;
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        return file;
    }
    /**
     * Download indices bhavcopy
     */
    async downloadIndicesBhavcopy(date, folder) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        const ddmmyyyy = (0, date_formatter_js_1.formatDateDDMMYYYY)(date);
        const url = `${index_js_1.ARCHIVE_URL}/content/indices/ind_close_all_${ddmmyyyy}.csv`;
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        return file;
    }
    /**
     * Download F&O bhavcopy
     */
    async downloadFnoBhavcopy(date, folder) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        const ymd = (0, date_formatter_js_1.formatDateYMD)(date);
        const url = `${index_js_1.ARCHIVE_URL}/content/fo/BhavCopy_NSE_FO_0_0_0_${ymd}_F_0000.csv.zip`;
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        return await (0, file_operations_js_1.unzipFile)(file, path_1.default.dirname(file));
    }
    /**
     * Download priceband report
     */
    async downloadPricebandReport(date, folder) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        const ddmmyyyy = (0, date_formatter_js_1.formatDateDDMMYYYY)(date);
        const url = `${index_js_1.ARCHIVE_URL}/content/equities/sec_list_${ddmmyyyy}.csv`;
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        return file;
    }
    /**
     * Download PR bhavcopy
     */
    async downloadPrBhavcopy(date, folder) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        const ddmmyy = (0, date_formatter_js_1.formatDateDDMMYY)(date);
        const url = `${index_js_1.ARCHIVE_URL}/archives/equities/bhavcopy/pr/PR${ddmmyy}.zip`;
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        return file;
    }
    /**
     * Download CM MII security report
     */
    async downloadCmMiiSecurityReport(date, folder) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        const ddmmyyyy = (0, date_formatter_js_1.formatDateDDMMYYYY)(date);
        const url = `${index_js_1.ARCHIVE_URL}/content/cm/NSE_CM_security_${ddmmyyyy}.csv.gz`;
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        return await (0, file_operations_js_1.unzipFile)(file, path_1.default.dirname(file));
    }
    /**
     * Download any document from URL
     */
    async downloadDocument(url, folder, extractFiles) {
        const outDir = folder ? (0, file_operations_js_1.getPath)(folder, true) : this.downloadDir;
        const file = await this.httpClient.downloadFile(url, outDir);
        if (!fs_1.default.existsSync(file)) {
            fs_1.default.rmSync(file, { force: true });
            throw new Error(`Failed to download file: ${path_1.default.basename(file)}`);
        }
        if (path_1.default.extname(file).toLowerCase() === ".zip") {
            try {
                return await (0, file_operations_js_1.unzipFile)(file, path_1.default.dirname(file), extractFiles);
            }
            catch (e) {
                fs_1.default.rmSync(file, { force: true });
                throw new Error(`Failed to extract zip file: ${String(e?.message ?? e)}`);
            }
        }
        return file;
    }
}
exports.DownloadApi = DownloadApi;
