// nse-bse-api/dist-cjs/nse/api/equity-api.js
"use strict";
/**
 * Equity-related API methods for NSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EquityApi = void 0;
const index_js_1 = require("../constants/index.js");
class EquityApi {
    httpClient;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * Get equity metadata information
     */
    async getEquityMetaInfo(symbol) {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/equity-meta-info`, {
            symbol: symbol.toUpperCase()
        });
    }
    /**
     * Get quote for equity or derivative
     */
    async getQuote(params) {
        const type = params.type ?? "equity";
        const url = type === "equity"
            ? `${index_js_1.BASE_URL}/quote-equity`
            : `${index_js_1.BASE_URL}/quote-derivative`;
        const query = { symbol: params.symbol.toUpperCase() };
        if (params.section) {
            if (params.section !== "trade_info") {
                throw new Error("'Section' if specified must be 'trade_info'");
            }
            query.section = params.section;
        }
        return await this.httpClient.request(url, query);
    }
    /**
     * Get simplified equity quote with OHLC data
     */
    async getEquityQuote(symbol) {
        const q = await this.getQuote({ symbol, type: "equity" });
        const v = await this.getQuote({ symbol, type: "equity", section: "trade_info" });
        const open = q.priceInfo?.open;
        const minmax = q.priceInfo?.intraDayHighLow;
        const close = q.priceInfo?.close;
        const ltp = q.priceInfo?.lastPrice;
        return {
            date: q.metadata?.lastUpdateTime,
            open,
            high: minmax?.max,
            low: minmax?.min,
            close: close ?? ltp,
            volume: v.securityWiseDP?.quantityTraded,
        };
    }
    /**
     * Filter gainers from market data
     */
    getGainers(data, count) {
        const filtered = data.data
            .filter((d) => d.pChange > 0)
            .sort((a, b) => b.pChange - a.pChange);
        return typeof count === "number" ? filtered.slice(0, count) : filtered;
    }
    /**
     * Filter losers from market data
     */
    getLosers(data, count) {
        const filtered = data.data
            .filter((d) => d.pChange < 0)
            .sort((a, b) => a.pChange - b.pChange);
        return typeof count === "number" ? filtered.slice(0, count) : filtered;
    }
    /**
     * List equity stocks by index
     */
    async listEquityStocksByIndex(index = "NIFTY 50") {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/equity-stockIndices`, { index });
    }
    /**
     * List all indices
     */
    async listIndices() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/allIndices`);
    }
    /**
     * List ETFs
     */
    async listEtf() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/etf`);
    }
    /**
     * List SME stocks
     */
    async listSme() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/live-analysis-emerge`);
    }
    /**
     * List Sovereign Gold Bonds
     */
    async listSgb() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/sovereign-gold-bonds`);
    }
}
exports.EquityApi = EquityApi;
