// nse-bse-api/dist-cjs/nse/api/index.js
"use strict";
/**
 * API modules for NSE client
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadApi = exports.MarketApi = exports.IpoApi = exports.CorporateApi = exports.HistoricalApi = exports.OptionsApi = exports.EquityApi = void 0;
var equity_api_js_1 = require("./equity-api.js");
Object.defineProperty(exports, "EquityApi", { enumerable: true, get: function () { return equity_api_js_1.EquityApi; } });
var options_api_js_1 = require("./options-api.js");
Object.defineProperty(exports, "OptionsApi", { enumerable: true, get: function () { return options_api_js_1.OptionsApi; } });
var historical_api_js_1 = require("./historical-api.js");
Object.defineProperty(exports, "HistoricalApi", { enumerable: true, get: function () { return historical_api_js_1.HistoricalApi; } });
var corporate_api_js_1 = require("./corporate-api.js");
Object.defineProperty(exports, "CorporateApi", { enumerable: true, get: function () { return corporate_api_js_1.CorporateApi; } });
var ipo_api_js_1 = require("./ipo-api.js");
Object.defineProperty(exports, "IpoApi", { enumerable: true, get: function () { return ipo_api_js_1.IpoApi; } });
var market_api_js_1 = require("./market-api.js");
Object.defineProperty(exports, "MarketApi", { enumerable: true, get: function () { return market_api_js_1.MarketApi; } });
var download_api_js_1 = require("./download-api.js");
Object.defineProperty(exports, "DownloadApi", { enumerable: true, get: function () { return download_api_js_1.DownloadApi; } });
