// nse-bse-api/dist-cjs/nse/api/ipo-api.js
"use strict";
/**
 * IPO-related API methods for NSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IpoApi = void 0;
const index_js_1 = require("../constants/index.js");
const date_formatter_js_1 = require("../utils/date-formatter.js");
class IpoApi {
    httpClient;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * List current IPOs
     */
    async listCurrentIPO() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/ipo-current-issue`);
    }
    /**
     * List upcoming IPOs
     */
    async listUpcomingIPO() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/all-upcoming-issues?category=ipo`);
    }
    /**
     * Get IPO details
     */
    async getIpoDetails(params) {
        const symbol = params.symbol.toUpperCase();
        const series = params.series ?? "EQ";
        return await this.httpClient.request(`${index_js_1.BASE_URL}/ipo-detail`, { symbol, series });
    }
    /**
     * List past IPOs
     */
    async listPastIPO(from_date, to_date) {
        const to = to_date ?? new Date();
        const from = from_date ?? new Date(to.getTime() - 90 * 86400000);
        if (to < from) {
            throw new Error("Argument `to_date` cannot be less than `from_date`");
        }
        return await this.httpClient.request(`${index_js_1.BASE_URL}/public-past-issues`, {
            from_date: (0, date_formatter_js_1.formatDateDMY)(from),
            to_date: (0, date_formatter_js_1.formatDateDMY)(to),
        });
    }
}
exports.IpoApi = IpoApi;
