// nse-bse-api/dist-cjs/nse/api/market-api.js
"use strict";
/**
 * Market data and trading API methods for NSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarketApi = void 0;
const index_js_1 = require("../constants/index.js");
const date_formatter_js_1 = require("../utils/date-formatter.js");
class MarketApi {
    httpClient;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * Get market status
     */
    async getStatus() {
        const result = await this.httpClient.request(`${index_js_1.BASE_URL}/marketStatus`);
        return result.marketState;
    }
    /**
     * Search/lookup symbols
     */
    async lookup(query) {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/search/autocomplete`, { q: query });
    }
    /**
     * Get block deals
     */
    async getBlockDeals() {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/block-deal`);
    }
    /**
     * Get bulk deals for date range
     */
    async getBulkDeals(fromDate, toDate) {
        const daysDiff = (toDate.getTime() - fromDate.getTime()) / 86400000;
        if (daysDiff > 365) {
            throw new Error("The date range cannot exceed one year.");
        }
        const url = `${index_js_1.BASE_URL}/historical/bulk-deals?from=${(0, date_formatter_js_1.formatDateDMY)(fromDate)}&to=${(0, date_formatter_js_1.formatDateDMY)(toDate)}`;
        const data = await this.httpClient.request(url);
        if (!data.data || data.data.length < 1) {
            throw new Error("No bulk deals data available for the specified date range.");
        }
        return data.data;
    }
    /**
     * Get holidays
     */
    async getHolidays(type = "trading") {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/holiday-master`, { type });
    }
    /**
     * Get daily reports file metadata
     */
    async getDailyReportsFileMetadata(segment = "CM") {
        return await this.httpClient.request(`${index_js_1.BASE_URL}/daily-reports`, { key: segment });
    }
}
exports.MarketApi = MarketApi;
