// nse-bse-api/dist-cjs/nse/index.js
"use strict";
/**
 * NSE API Module
 *
 * Re-exports all NSE functionality from the original nseapi package
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpClient = exports.DownloadApi = exports.MarketApi = exports.IpoApi = exports.CorporateApi = exports.HistoricalApi = exports.OptionsApi = exports.EquityApi = exports.NSEClient = void 0;
// Main NSE client
var nse_client_js_1 = require("./nse/nse-client.js");
Object.defineProperty(exports, "NSEClient", { enumerable: true, get: function () { return nse_client_js_1.NSEClient; } });
// Export types for TypeScript users
__exportStar(require("./types/index.js"), exports);
// Export constants
__exportStar(require("./constants/index.js"), exports);
// Export utilities
__exportStar(require("./utils/date-formatter.js"), exports);
__exportStar(require("./utils/file-operations.js"), exports);
// Export API modules for advanced usage
var equity_api_js_1 = require("./api/equity-api.js");
Object.defineProperty(exports, "EquityApi", { enumerable: true, get: function () { return equity_api_js_1.EquityApi; } });
var options_api_js_1 = require("./api/options-api.js");
Object.defineProperty(exports, "OptionsApi", { enumerable: true, get: function () { return options_api_js_1.OptionsApi; } });
var historical_api_js_1 = require("./api/historical-api.js");
Object.defineProperty(exports, "HistoricalApi", { enumerable: true, get: function () { return historical_api_js_1.HistoricalApi; } });
var corporate_api_js_1 = require("./api/corporate-api.js");
Object.defineProperty(exports, "CorporateApi", { enumerable: true, get: function () { return corporate_api_js_1.CorporateApi; } });
var ipo_api_js_1 = require("./api/ipo-api.js");
Object.defineProperty(exports, "IpoApi", { enumerable: true, get: function () { return ipo_api_js_1.IpoApi; } });
var market_api_js_1 = require("./api/market-api.js");
Object.defineProperty(exports, "MarketApi", { enumerable: true, get: function () { return market_api_js_1.MarketApi; } });
var download_api_js_1 = require("./api/download-api.js");
Object.defineProperty(exports, "DownloadApi", { enumerable: true, get: function () { return download_api_js_1.DownloadApi; } });
var http_client_js_1 = require("./http/http-client.js");
Object.defineProperty(exports, "HttpClient", { enumerable: true, get: function () { return http_client_js_1.HttpClient; } });
